log/loki.go

```diff
@@ -0,0 +1,422 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package log
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// TODO move this to it's own package
+// benchmark it
+
+type lokiHook struct {
+	addr             string
+	additionalParams [][2]string
+	ch               chan *logrus.Entry
+	limit            int
+	levels           []logrus.Level
+	pushPeriod       time.Duration
+	client           *http.Client
+}
+
+// LokiFromConfigLine returns a new logrus.Hook that pushes logrus.Entrys to loki and is configured
+// through the provided line
+func LokiFromConfigLine(line string) (logrus.Hook, error) {
+	h := &lokiHook{
+		addr:       "http://127.0.0.1:3100/loki/api/v1/push",
+		limit:      100,
+		levels:     logrus.AllLevels,
+		pushPeriod: time.Second * 1, // TODO configurable,
+		client: &http.Client{
+			Timeout: time.Second * 5, // TODO configurable, equal to pushPeriod?
+		},
+	}
+	if line == "loki" {
+		return h, nil
+	}
+
+	parts := strings.SplitN(line, "=", 2)
+	if parts[0] != "loki" {
+		return nil, fmt.Errorf("loki configuration should be in the form `loki=url-to-push` but is `%s`", line)
+	}
+	args := strings.Split(parts[1], ",")
+	h.addr = args[0]
+	// TODO use something better ... maybe
+	// https://godoc.org/github.com/kubernetes/helm/pkg/strvals
+	// atleast until https://github.com/loadimpact/k6/issues/926?
+	if len(args) == 1 {
+		return h, nil
+	}
+
+	for _, arg := range args[1:] {
+		paramParts := strings.SplitN(arg, "=", 2)
+
+		if len(paramParts) != 2 {
+			return nil, fmt.Errorf("loki arguments should be in the form `address,key1=value1,key2=value2`, got %s", arg)
+		}
+
+		key := paramParts[0]
+		value := paramParts[1]
+		switch key {
+		case "additionalParams":
+			values := strings.Split(value, ";") // ; because , is already used
+
+			h.additionalParams = make([][2]string, len(values))
+			for i, value := range values {
+				additionalParamParts := strings.SplitN(value, "=", 2)
+				if len(additionalParamParts) != 2 {
+					return nil, fmt.Errorf("additionalparam should be in the form key1=value1;key2=value2, got %s", value)
+				}
+				h.additionalParams[i] = [2]string{additionalParamParts[0], additionalParamParts[1]}
+			}
+		case "limit":
+			var err error
+			h.limit, err = strconv.Atoi(value)
+			if err != nil {
+				return nil, fmt.Errorf("couldn't parse the loki limit as a number %w", err)
+			}
+		case "level":
+			// TODO figure out if `tracing`,`fatal` and `panic` should be included
+			h.levels = []logrus.Level{}
+			switch value {
+			case "debug":
+				h.levels = append(h.levels, logrus.DebugLevel)
+				fallthrough
```

Comment:
Instead of `fallthrough`, it will be cleaner if you have a map between the string and logrus levels, like this: https://github.com/loadimpact/k6/blob/1fcab52ce5faf914aa3360c1c07c207be4a2f94e/js/console_test.go#L97-L103
And then use the [`logrus.AllLevels`](https://godoc.org/github.com/sirupsen/logrus#pkg-variables) variable to iterate and append all "lower" levels to whatever was configured in `h.levels`?

Reply:
I find the switch to be much more elegant than going through some slice that may or may not be ordered , at best I can change it to be  a switch which just direactly sets the final slice as:
```
case "info":
     h.levels = []logrus.Level{logrus.InfoLevel, logrus.WarnLevl, logrus.ErrorLevel}
```

But I find the current code to be ... pretty understandable and clean ... :D

Reply:
It's pretty obvious that constant is ordered... And if not, you can define the same slice yourself, to make extra sure it's ordered... :wink: I don't really care, though it will still be better to take this out as a short function somewhere